boa/src/syntax/lexer/cursor.rs

```diff
@@ -70,31 +69,94 @@ where
                 true
             }
             _ => false,
-            Some(&Err(_)) => return self.peeked.take().unwrap().unwrap().map(|_| false),
+            // Some(&Err(_)) => return self.peeked.take().unwrap().unwrap().map(|_| false),
+        })
+    }
+
+    /// Applies the predicate to the next character and returns the result.
+    /// Returns false if there is no next character.
+    ///
+    /// The buffer is not incremented.
+    #[inline]
+    pub(super) fn next_is_pred<F>(&mut self, pred: &F) -> io::Result<bool>
+    where
+        F: Fn(char) -> bool,
+    {
+        Ok(match self.peek() {
+            None => false,
+            Some(Ok(peek)) => pred(*peek),
+            Some(Err(e)) => todo!(),
         })
     }
 
     /// Fills the buffer with all characters until the stop character is found.
     ///
     /// Note: It will not add the stop character to the buffer.
     pub(super) fn take_until(&mut self, stop: char, buf: &mut String) -> io::Result<()> {
-        unimplemented!()
+        loop {
+            if self.next_is(stop)? {
+                return Ok(());
+            } else {
+                match self.next() {
+                    None => {
+                        return Err(io::Error::new(
+                            ErrorKind::UnexpectedEof,
+                            format!("Unexpected end of file when looking for character {}", stop),
+                        ));
+                    }
+                    Some(Err(e)) => {
+                        return Err(e);
+                    }
+                    Some(Ok(ch)) => {
+                        buf.push(ch);
+                    }
+                }
+            }
+        }
+    }
+
+    /// Fills the buffer with characters until the first character (x) for which the predicate (pred) is false
+    /// (or the next character is none).
+    ///
+    /// Note that all characters up until x are added to the buffer including the character right before.
+    pub(super) fn take_until_pred<F>(&mut self, buf: &mut String, pred: &F) -> io::Result<()>
+    where
+        F: Fn(char) -> bool,
+    {
+        loop {
+            if !self.next_is_pred(pred)? {
+                return Ok(());
+            } else {
+                match self.next() {
+                    None => {
+                        unimplemented!();
+                        // unreachable!();
+                    }
+                    Some(Err(e)) => {
+                        return Err(e);
+                    }
+                    Some(Ok(ch)) => {
+                        buf.push(ch);
+                    }
+                }
+            }
+        }
     }
 
     /// Retrieves the given number of characters and adds them to the buffer.
-    pub(super) fn take(&mut self, count: usize, buf: &mut String) -> io::Result<()> {
+    pub(super) fn _take(&mut self, _count: usize, _buf: &mut String) -> io::Result<()> {
```

Comment:
There is no need to replace all these with an underscore (especially the function name, should never start with an underscore).

Warnings are expected while in development.

Reply:
Will revert before marking branch as ready, how come function names should never start with an underscore? (I thought it was just used to mark a function as unused currently?)

Reply:
Well, a function should be used or not present. If we are in the process of developing a new feature, then it's understandable that we get "unused" warnings, but we should not silence them, so that we know where to look before moving code to the master branch.

Reply:
Fixed